lib/Connection.js

```diff
@@ -134,49 +166,46 @@ class Connection {
    * Invokes transport with XML input from command list
    * Calls user provided callback with the XML output
    * @param {fuction} callback
-   * @param {boolean} sync
    */
   run(callback) {
+    if (typeof callback !== 'function') {
+      throw new Error('Passing a callback function is required');
+    }
     // Build the XML document.
-    let xml = iXml.iXmlNodeHead() + iXml.iXmlNodeScriptOpen();
+    let xmlInput = iXmlNodeHead() + iXmlNodeScriptOpen();
+
     for (let i = 0; i < this.cmds.length; i += 1) {
-      xml += this.cmds[i];
+      xmlInput += this.cmds[i];
     }
-    xml += iXml.iXmlNodeScriptClose();
-    this.cmds = [];
+
+    xmlInput += iXmlNodeScriptClose();
+    this.cmds = []; // reset the commands list
+
     // Print the XML document if in debug mode.
-    if (this.I_DEBUG_VERBOSE) {
+    if (this.conn.verbose) {
       console.log('============\nINPUT XML\n============');
-      console.log(xml);
+      console.log(xmlInput);
       console.log('============\nOUTPUT XML\n============');
     }
+
     // Post the XML document to XML service program.
-    if (this.conn.I_TRANSPORT === I_TRANSPORT_REST) {
-      iRestHttp(callback,
-        this.conn.I_TRANSPORT_REST_HOST,
-        this.conn.I_TRANSPORT_REST_PORT,
-        this.conn.I_TRANSPORT_REST_PATH,
-        this.conn.I_TRANSPORT_DB2_DATABASE,
-        this.conn.I_TRANSPORT_DB2_USER,
-        this.conn.I_TRANSPORT_DB2_PASSWORD,
-        this.conn.I_TRANSPORT_IPC,
-        this.conn.I_TRANSPORT_CTL,
-        xml,
-        this.conn.I_TRANSPORT_REST_XML_OUT_SIZE);
-    } else if (this.conn.I_TRANSPORT === I_TRANSPORT_DB2) {
-      db2Call(callback,
-        this.conn.I_XML_SERVICE_LIB,
-        this.conn.I_TRANSPORT_DB2_DATABASE,
-        this.conn.I_TRANSPORT_DB2_USER,
-        this.conn.I_TRANSPORT_DB2_PASSWORD,
-        this.conn.I_TRANSPORT_IPC,
-        this.conn.I_TRANSPORT_CTL,
-        xml,
-        this.conn.I_TRANSPORT_REST_XML_OUT_SIZE,
-        this.I_DEBUG_VERBOSE);
-    } else {
-      callback(iXml.iXmlNodeHead() + iXml.iXmlNodeScriptOpen() + iXml.iXmlNodeError('***Transport error no data***') + iXml.iXmlNodeScriptClose());
-    }
+    this.conn.transportCall(this.conn, xmlInput, (error, xmlOut) => {
+      // Create an Error object if error returned from transport is not one already
+      if (error && !(error instanceof Error)) {
+        // eslint-disable-next-line no-param-reassign
+        error = new Error(error);
+      }
+
+      if (this.conn.returnError) {
+        callback(error, xmlOut);
+      } else {
+        // when return return error is false, let it be known!
+        if (error) {
+          throw error;
```

Comment:
Won't this break existing applications? How does this even work? ie. Where does the exception get thrown to?

Reply:
Yeah good call I will remove this. If you choose to turn off `returnError` then you will be blind to the error.